etcdwatcher/etcdwatcher_test.go

```diff
@@ -0,0 +1,2025 @@
+
+	"github.com/rudderlabs/rudder-go-kit/jsonrs"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+
+	"github.com/rudderlabs/rudder-go-kit/etcdwatcher"
+)
+
+// TestData represents the structure of data stored in etcd for testing
+type TestData struct {
+	ID   int    `json:"id"`
+	Name string `json:"name"`
+}
+
+// MockEtcdClient is a mock implementation of the etcd client
+type MockEtcdClient struct {
+	watchChans map[string]chan clientv3.WatchResponse
+	getResp    *clientv3.GetResponse
+	getErr     error
+}
+
+func (m *MockEtcdClient) Get(_ context.Context, _ string, _ ...clientv3.OpOption) (*clientv3.GetResponse, error) {
+	return m.getResp, m.getErr
+}
+
+func (m *MockEtcdClient) Watch(_ context.Context, key string, _ ...clientv3.OpOption) clientv3.WatchChan {
+	return m.watchChans[key]
+}
+
+// Constructor tests
+func TestWatcher_New(t *testing.T) {
+	t.Run("valid construction with default values", func(t *testing.T) {
+		client := &MockEtcdClient{}
+		watcher, err := etcdwatcher.New[TestData](client, "/test/key")
+
+		require.NoError(t, err)
+		require.NotNil(t, watcher)
+	})
+
+	t.Run("valid construction with explicit values", func(t *testing.T) {
+		client := &MockEtcdClient{}
+		watcher, err := etcdwatcher.New[TestData](client, "/test/",
+			etcdwatcher.WithPrefix[TestData](),
+			etcdwatcher.WithWatchEventType[TestData](etcdwatcher.PutWatchEventType),
+			etcdwatcher.WithWatchMode[TestData](etcdwatcher.AllMode))
+
+		require.NoError(t, err)
+		require.NotNil(t, watcher)
+	})
+
+	t.Run("error when using OnceMode with non-prefix watches", func(t *testing.T) {
+		client := &MockEtcdClient{}
+		watcher, err := etcdwatcher.New[TestData](client, "/test/key",
+			etcdwatcher.WithWatchMode[TestData](etcdwatcher.OnceMode))
+
+		require.Error(t, err)
+		require.Nil(t, watcher)
+		require.Contains(t, err.Error(), "once mode can only be used with prefix watches")
+	})
+
+	t.Run("valid construction with OnceMode and prefix watches", func(t *testing.T) {
+		client := &MockEtcdClient{}
+		watcher, err := etcdwatcher.New[TestData](client, "/test/",
+			etcdwatcher.WithPrefix[TestData](),
+			etcdwatcher.WithWatchMode[TestData](etcdwatcher.OnceMode))
+
+		require.NoError(t, err)
+		require.NotNil(t, watcher)
+	})
+}
+
+// TestLoadAndWatch_AllMode_AllEventType tests LoadAndWatch with AllMode and AllWatchEventType
+func TestLoadAndWatch_AllMode_AllEventType(t *testing.T) {
+	pool, err := dockertest.NewPool("")
+	require.NoError(t, err)
+
+	resource, err := etcd.Setup(pool, t)
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	// Prepare test data
+	testData := TestData{ID: 1, Name: "initial"}
+	dataBytes, err := jsonrs.Marshal(testData)
+	require.NoError(t, err)
+
+	// Put initial data
+	_, err = resource.Client.Put(ctx, "/test/key1", string(dataBytes))
+	require.NoError(t, err)
+
+	// Create watcher
+	watcher, err := etcdwatcher.New[TestData](resource.Client, "/test/",
```

Comment:
can we add this before each new watcher we're creating?
```go
defer goleak.VerifyNone(t, goleak.IgnoreCurrent())
```

I guess we should also cancelWatch at the end and wait, e.g.

```go
	// leave
	cancelWatch()
	select {
	case _, ok := <-eventCh:
		require.False(t, ok, "Expected channel to be closed")
	case <-time.After(10 * time.Second):
		t.Fatal("Did not receive expected DELETE event")
	}
```